### ISSUE TITLE
Increase PPO Entropy Coefficient with Scheduled Decay

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#50]
## Problem

PPO's current entropy coefficient (`0.01`) provides only ~2,000-5,000 exploration steps for a 19-action space. This is **10x insufficient** (need ~19,000 steps).

**Result**: Agent finds Action 6 early and never explores alternatives, achieving 87-90% single-action dominance.

---

## Proposed Solution: Increase Entropy Coefficient

**Simple fix**: Increase initial entropy coefficient and decay over training.

### Current Configuration
```python
entropy_coef = 0.01  # Fixed throughout training
```

**Problems:**
- Provides only 0.01 × reward_magnitude incentive for exploration
- Insufficient for complex action spaces
- Agent exploits prematurely

### Proposed Configuration

```python
class EntropyScheduler:
    """Decay entropy coefficient during training."""
    
    def __init__(
        self,
        initial_coef: float = 0.1,
        final_coef: float = 0.01,
        decay_frames: int = 80_000,
        decay_type: str = "linear"
    ):
        self.initial = initial_coef
        self.final = final_coef
        self.decay_frames = decay_frames
        self.decay_type = decay_type
    
    def get_coef(self, current_frame: int) -> float:
        if current_frame >= self.decay_frames:
            return self.final
        
        progress = current_frame / self.decay_frames
        
        if self.decay_type == "linear":
            return self.initial + (self.final - self.initial) * progress
        elif self.decay_type == "exponential":
            return self.final + (self.initial - self.final) * (0.5 ** (progress * 5))
        elif self.decay_type == "cosine":
            return self.final + (self.initial - self.final) * 0.5 * (1 + np.cos(np.pi * progress))
        
        return self.initial
```

---

### Option A: Linear Decay (Recommended)
```
Frames 0-80k:     0.1 → 0.01 (linear)
Frames 80k-100k:  0.01 (constant)
```
- **Exploration**: First 80% of training
- **Exploitation**: Final 20% to converge

### Option B: Exponential Decay
```
Frames 0-40k:     0.1 → 0.03 (fast decay)
Frames 40k-80k:   0.03 → 0.01 (slow decay)
Frames 80k-100k:  0.01 (constant)
```
- Front-loads exploration
- Smoother transition to exploitation

### Option C: Cosine Annealing
```
Frames 0-80k:     0.1 → 0.01 (cosine curve)
Frames 80k-100k:  0.01 (constant)
```
- Smooth decay following cosine curve
- Popular in deep learning optimization

---

### Exploration Budget Increase

```
Current (coef=0.01):
  Exploration steps ≈ 2,000-5,000
  Per-action samples ≈ 100-250

Proposed (coef=0.1 → 0.01):
  Exploration steps ≈ 15,000-20,000
  Per-action samples ≈ 800-1,000
```

### Success Metrics

| Metric | Baseline | Target |
|--------|----------|--------|
| Action entropy | 0.48 | > 1.0 |
| Actions with >5% usage | 1-2 | > 8 |
| Action 6 dominance | 87% | < 50% |
| Exploration phase duration | 10k frames | 40-60k frames |

---

## Integration

```python
# In train.py
entropy_scheduler = EntropyScheduler(
    initial_coef=0.1,
    final_coef=0.01,
    decay_frames=80_000,
    decay_type="linear"
)

# In training loop
for i, batch in enumerate(collector):
    current_entropy_coef = entropy_scheduler.get_coef(collected_frames)
    
    # Update loss module
    ppo_loss.entropy_coef = current_entropy_coef
    
    # Log for monitoring
    logger.log_scalar("train/entropy_coef", current_entropy_coef, collected_frames)
```

---

### Week 1: Core Implementation
- [ ] Create `EntropyScheduler` in `torchtrade/training/schedulers.py`
- [ ] Add `entropy_schedule` config to training configs
- [ ] Integrate with PPO loss module
- [ ] Add logging of current entropy coefficient

### Week 2: Experiments
- [ ] Baseline: Fixed `coef=0.01` (reproduce current behavior)
- [ ] Test 1: Fixed `coef=0.1` (no decay)
- [ ] Test 2: Linear decay `0.1 → 0.01` over 80k frames
- [ ] Test 3: Exponential decay
- [ ] Test 4: Cosine annealing

### Week 3: Analysis & Optimization
- [ ] Compare action distributions across experiments
- [ ] Analyze convergence speed vs exploration thoroughness
- [ ] Find optimal initial_coef and decay_frames
- [ ] A/B test vs curriculum learning (#49)

---

## Configuration Example

```yaml
# config.yaml
loss:
  entropy_coef: 0.1  # Initial value
  entropy_schedule:
    enabled: true
    decay_type: "linear"  # or "exponential", "cosine"
    final_coef: 0.01
    decay_frames: 80000
  
  # Existing params
  gamma: 0.9
  clip_epsilon: 0.1
  # ...
```

---

## Advantages vs Other Approaches

**vs Curiosity (#48):**
- ✅ Simpler to implement (no new modules)
- ✅ No additional hyperparameters to tune
- ❌ Less targeted (explores randomly, not intelligently)

**vs Curriculum (#49):**
- ✅ No action masking logic needed
- ✅ Agent can explore full space from start
- ❌ Less structured exploration

**Best approach:** Combine with curriculum or curiosity for maximum effect!

---

## Monitoring & Debugging

Track these metrics during training:

```python
metrics = {
    "train/entropy_coef": current_coef,
    "train/policy_entropy": policy.dist.entropy().mean(),
    "train/action_std": actions.float().std(),
    "train/unique_actions_10k": len(set(actions[-10000:])),
}
```

**Expected behavior:**
- High entropy coefficient → High policy entropy
- Policy entropy should track coefficient decay
- Unique actions should increase during exploration phase

---

## References

**Entropy in RL:**
- Williams & Peng (1991) - "Function optimization using connectionist reinforcement learning algorithms"
- Mnih et al. (2016) - "Asynchronous Methods for Deep Reinforcement Learning" (A3C with entropy)

**Scheduling:**
- Loshchilov & Hutter (2017) - "SGDR: Stochastic Gradient Descent with Warm Restarts" (cosine annealing)

---

## Related Issues

- #48 - Curiosity-driven exploration (can be combined)
- #49 - Curriculum learning (can be combined)
- Related: `RewardFunctionResearch/FuturesOneStepEnv_Iteration5/reports/SUMMARY_iterations_5-10.md`
